src/mlpack/core/tree/greedy_single_tree_traverser_impl.hpp

```diff
@@ -24,26 +24,48 @@ template<typename TreeType, typename RuleType>
 GreedySingleTreeTraverser<TreeType, RuleType>::GreedySingleTreeTraverser(
     RuleType& rule) :
     rule(rule),
-    numPrunes(0)
+    numPrunes(0),
+    k(0)
 { /* Nothing to do. */ }
 
 template<typename TreeType, typename RuleType>
 void GreedySingleTreeTraverser<TreeType, RuleType>::Traverse(
     const size_t queryIndex,
     TreeType& referenceNode)
 {
-  // Run the base case as necessary for all the points in the reference node.
-  for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
-    rule.BaseCase(queryIndex, referenceNode.Point(i));
+  if(referenceNode.IsLeaf())
+  {
```

Comment:
So, this conditional actually is not necessary.  While the kd-tree only holds points in the leaves, other trees (like the cover tree) may hold some points in non-leaf nodes.  So we should calculate the base case of any points even as we go down.  We may have to track the total number of base cases computed in the traverser, but I think that is just fine to do.

Reply:
Right! Imposing this condition is just a reduction of results for trees storing points in other nodes also. This will just make them behave as the earlier greedy kd-tree.